sql: set running status for GC jobs

postamar opened the pull request:
Previously GC jobs didn't populate their RunningStatus leaving users no
way to tell whether a job is waiting on the timer or actually doing
garbage collection. This change addresses this.

Fixes #57826.

Release note (sql change): GC jobs now populate the running_status
column for SHOW JOBS.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58612)
<!-- Reviewable:end -->

thoszhang reviewed (commented):
> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58612)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[pkg/sql/gcjob/gc_job.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58612#-MQZJpr49H0dbJSV52P5:-MQZHdj4CQqv2DmSlbtV:b63q278) ([raw file](https://github.com/cockroachdb/cockroach/blob/5869e47c48f127a5d0c9fca709f22bc333545c1c/pkg/sql/gcjob/gc_job.go#L36)):*
> ```Go
> 
> // runningStatusPerformGC is for jobs that are currently garbage-collecting.
> const runningStatusPerformGC jobs.RunningStatus = "garbage-collecting"
> ```

Maybe `performing garbage collection`?
___
*[pkg/sql/gcjob/gc_job.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58612#-MQZJhjz9-U4V0wD3yPx:-MQZJhjz9-U4V0wD3yPy:b-ai6eh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5869e47c48f127a5d0c9fca709f22bc333545c1c/pkg/sql/gcjob/gc_job.go#L147)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> 			// Some elements have been marked as DELETING so save the progress.
> 			persistProgress(ctx, execCfg, r.jobID, progress, runningStatusPerformGC)
> 			if err := performGC(ctx, execCfg, details, progress); err != nil {
> 				return err
> 			}
> 			persistProgress(ctx, execCfg, r.jobID, progress, sql.RunningStatusWaitingGC)
> ```
</details>


So the behavior here is that the status is `runningStatusPerformGC` when we're actively performing GC, and then gets reset to `RunningStatusWaitingGC` afterward if there are more tables to wait for, right? Thinking about this more, I think it would be helpful to indicate which table IDs are being GCed when we display this, or else it might be confusing when the status switches back to `RunningStatusWaitingGC`. But we'd have to refactor a bit to return that information from `refreshTables`, I think.

In any case, we'd want to test this behavior. I think the way to do that is to add some testing knobs (like in `GCJobTestingKnobs` and `SchemaChangerTestingKnobs`) to have the schema change job block or get restarted right after we write the `runningStatusPerformGC` status, and then query for the job status in that state.

Honestly, I'd be fine with just always having the job status be `waiting for GC TTL` for now, since that's how it was pre-20.1. The other enhancements might not be worth making at this point. If we backport this change, we'd certainly want to only backport that limited change.


<!-- Sent from Reviewable.io -->

postamar commented:
I pushed an amended commit based on your comments, going for the limited change you suggested, and adding tests & the testing knobs (did I do this right?). Adding the table IDs to the running status message looks easy enough but the devil's in the details. Conceivably it's possible to perform GC on, say, a thousand tables, right? Should we print a thousand table IDs in the running status message in that case? In any case I'm quite happy to submit another PR once this one is merged and backported.

I'm assuming, based on your comments, that pre-20.1, the running status message was either `sql.RunningStatusWaitingGC` or NULL. Did I understand correctly? I tried figuring things out for myself by checking out `release-19.2` but things changed a lot since then so I figured I'd just ask.

thoszhang reviewed (commented):
> I'm assuming, based on your comments, that pre-20.1, the running status message was either `sql.RunningStatusWaitingGC` or NULL. Did I understand correctly?

No, I'm pretty sure it was just literally always `sql.RunningStatusWaitingGC` and never changed, even when GC was happening. (This was back when GC happened as part of the schema change job.) So, even simpler.

I would recommend just splitting off another PR from this one that just sets `sql.RunningStatusWaitingGC` as the status and never changes it. At that point, all you really need is to make a small change in the logic tests; there's no real point in adding a testing knob for just that change. I would be content with merging and backporting that, and then possibly revisiting more changes later.

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58612#-:-MR3_EiP0tjQWGUMa94N:blhw6de)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)



<!-- Sent from Reviewable.io -->

Commit:
sql: set running status for GC jobs

Previously GC jobs didn't populate their RunningStatus leaving users
confused. This change addresses this by setting it to
sql.RunningStatusWaitingGC, as it was pre-20.1

Fixes #57826.

Release note (bug fix): GC jobs now populate the running_status column
for SHOW JOBS. This bug has been present since v20.1.

postamar commented:
Sounds good, I've done just that. I'll open a backport PR if/when this gets approved.

thoszhang reviewed (approved):
> Reviewed 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58612)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

postamar commented:
bors r+
